### ISSUE TITLE
Message received in unrecognized characters

### ISSUE THREAD

**Issue by ivanlewis** [mgp25/Chat-API-NET#132]
Original message sent was 'Hi'

Received text was 
3�!�~i?��G#????.D�?????[ ??�?H?0�!�?i??B��+p_???Q?\?2???T?#�]?\ x�"B3
!�?????�hW???????]]C^?��2#???]?P���

what is the issue?

**Comment by yniv** [comment]
Yep... the c# API's doesn't support the axolotl encryption. WhatsApp finally rolled the change to make all numbers use encryption. SO, until chat-api-net or any other c# api will support it they are all useless.... :-/

**Comment by alijundi** [comment]
Yes I am experiencing the same issue!

Encryption is added! Is there a way to decrypt it?

**Comment by samuelg78** [comment]
i am experiencing the issue..... but for some numbers only.

**Comment by alijundi** [comment]
We should think about porting the decryption method same in php library.

**Comment by yniv** [comment]
@samuelg78 what old version??? there is a version that works good?

**Comment by samuelg78** [comment]
yes..... how can i upload to this web. I got the error "Unfortunately, we don’t support that file type.  Choose Files Try again with a PNG, GIF, JPG, DOCX, PPTX, XLSX, TXT, PDF, or ZIP."

My file is ZIP format.

**Comment by yniv** [comment]
@samuelg78 don't you have the code of that version??? what's in the WhatsConstants.cs file?

**Comment by samuelg78** [comment]
cos that file is done by freelancer... i dun have the code

**Comment by alijundi** [comment]
I think we can use the C# version of axolotl encryption at https://github.com/langboost/libaxolotl-uwp.

I just need some hints on how to get the public key, private key and when they are transmitted in order to proceed. Maybe @mgp25 can help us in this regard.

@mgp can you help us here.

Thanks.

**Comment by samuelg78** [comment]
Can anyone tell me which tools/program to use for sniffing the parameter sent to whatsapp server?

**Comment by alijundi** [comment]
you can use wireshark. download it at https://www.wireshark.org/#download

**Comment by samuelg78** [comment]
I had installed it.... but which data to look out for? Pls advice.

**Comment by rflsouza** [comment]
@alijundi, The wiki has a page that explains a little operation of E2E encryption. 

https://github.com/WHAnonymous/Chat-API/wiki/End-to-end-encryption I believe this can give help.

Best regards

**Comment by ricardoperovano** [comment]
@samuelg78 send me the whatsappapi.dll, i can see its code!!

**Comment by samuelg78** [comment]
This WhatsAppApi.dll version somehow still work for me. But i dun have the source code anymore. If anyone able to get version or able to reverse engineer. Let me know

http://xz1.mine.nu/WhatsAppApi.zip

**Comment by ricardoperovano** [comment]
I can see the code, what should I look for ?

**Comment by ricardoperovano** [comment]
public class WhatsConstants
	{
		public const string WhatsAppDigest = "xmpp/s.whatsapp.net";

		public const string WhatsAppHost = "c3.whatsapp.net";

		public const string WhatsAppRealm = "s.whatsapp.net";

		public const string WhatsAppServer = "s.whatsapp.net";

		public const string WhatsGroupChat = "g.us";

		public const string WhatsAppVer = "2.11.16";

		public const int WhatsPort = 443;

		public const string Device = "iPhone";

		public const string UserAgent = "WhatsApp/2.12.68 S40Version/14.26 Device/Nokia302";

		public static NumberStyles WhatsAppNumberStyle = NumberStyles.AllowLeadingSign | NumberStyles.AllowDecimalPoint;

		public static DateTime UnixEpoch = new DateTime(1970, 1, 1, 0, 0, 0, 0, DateTimeKind.Utc);
	}

**Comment by ricardoperovano** [comment]
@samuelg78 its working. i get the class whatsconstant and watoken from your dll and replace on mine and its working

**Comment by ricardoperovano** [comment]
i think that its for iphone, anyway, it solved the problem for now. lets see for how long :(

**Comment by ricardoperovano** [comment]
Also i replaced this code on my watoken
string s = "PdA2DJyKoUrwLw1Bg6EIhzh502dF9noR9uFCllGk1425519315543" + number;

**Comment by yniv** [comment]
@ricardoperovano thats all you changed in watoken?

**Comment by samuelg78** [comment]
How do you manage to see the code? Please advice.

**Comment by samuelg78** [comment]
@ricardoperovano Please post the code again. Thks

**Comment by yniv** [comment]
here's the code. can someone confirm it works and block free???

```
public const string WhatsAppHost = "c3.whatsapp.net";

    public const string WhatsAppRealm = "s.whatsapp.net";

    public const string WhatsAppServer = "s.whatsapp.net";

    public const string WhatsGroupChat = "g.us";

    public const string WhatsAppVer = "2.11.16";

    public const int WhatsPort = 443;

    public const string Device = "iPhone";

    public const string UserAgent = "WhatsApp/2.12.68 S40Version/14.26 Device/Nokia302";


Also i replaced this code on my watoken
string s = "PdA2DJyKoUrwLw1Bg6EIhzh502dF9noR9uFCllGk1425519315543" + number;
```

**Comment by arielschwindt** [comment]
It works.

Enviado desde mi iPhone

**Comment by mgp25** [comment]
Requires libaxolotl, i'll do it eventually

Regards
